### ISSUE TITLE
Azure SQL DacPac deployment failing when trying to drop a column even though /p:BlockOnPossibleDataLoss=false is set for SqlPackage.exe

### ISSUE THREAD

**Issue by usmanibnesadiq** [microsoft/azure-pipelines-tasks#11191]
Hi, my release in Azure DevOps fails when trying to deploy a dacpac saying:

"The column [dbo].[table_Name].[bRowVersion] is being dropped, data loss could occur."

I have already set the parameter to SqlPackage.exe to ignore this behavior. Also, why is this showing up as a warning and still being treated like an error, blocking the deployment.

![image](https://user-images.githubusercontent.com/5325072/63556620-e170c900-c588-11e9-9f60-1b70cee32c46.png)

**Comment by usmanibnesadiq** [comment]
Knowing that my changes would not break anything on target database, I could set /p:TreatVerificationErrorsAsWarnings=true and the deployment works. But this is a dangerous thing to do. There should be a proper workaround. This SOF thread has some answers but not convinced fully 

https://stackoverflow.com/questions/28618386/ssdt-publish-errors-on-creating-publish-preview

**Comment by caphuevo** [comment]
There is any progress with this issue?

**Comment by JosephG3001** [comment]
Setting /p:BlockOnPossibleDataLoss=true worked for me.  I can delete columns without the release pipeline failing.

**Comment by larsts** [comment]
@JosephG3001 : Did you mean /p:BlockOnPossibleDataLoss=_false_ ?
Setting /p:BlockOnPossibleDataLoss=false worked for me. 

Release without flag failed with "The column [xxx] is being dropped, data loss could occur". Updating pipeline adding /p:BlockOnPossibleDataLoss=false and creating new release succeeded.

**Comment by JosephG3001** [comment]
@larsts Yes sorry:  /p:BlockOnPossibleDataLoss=false

**Comment by 20shivangi** [comment]
Feel free to reopen the issue, if you face any.

**Comment by jeroenski74** [comment]
Still got this issue when using it in the Deploy Azure Data Warehouse Publish (which is just new) don't think it should matter anything since it is using SQLPackage underneath but hope you can help out:

Settings used:

/p:GenerateSmartDefaults=True /p:TreatVerificationErrorsAsWarnings=False /p:BlockOnPossibleDataLoss=False /p:AllowDropBlockingAssemblies=True /p:DropObjectsNotInSource=True /p:DoNotDropObjectTypes=Users,Logins,RoleMembership,ExternalDataSources,DatabaseScopedCredentials /p:ExcludeObjectTypes=Users;Logins;RoleMembership;ExternalDataSources;DatabaseScopedCredentials

Error:

2019-10-24T10:37:10.8080874Z ##[error]*** Could not deploy package.
2019-10-24T10:37:10.8121669Z ##[error]Warning SQL72015: The type for column ID in table [DM].[DimConcurrent] is currently  NVARCHAR (255) NOT NULL but is being changed to  NVARCHAR (50) NOT NULL. Data loss could occur.

**Comment by jeroenski74** [comment]
> Feel free to reopen the issue, if you face any.

Can you reopen this issue?

**Comment by eaarora-ms** [comment]
Hi,
I have forwarded the issue to the SQL team. We will let you know when we get an update from them.

**Comment by Benjin** [comment]
@usmanibnesadiq Are you still experiencing this issue?

**Comment by eaarora-ms** [comment]
@jeroenski74  Can you please confirm if you are still facing the issue? Thanks!

**Comment by sankara-narayanan** [comment]
I can confirm that the flag /p:BlockOnPossibleDataLoss=false is absolutely not honored. I have been facing similar issues. i can provide any further evidences you might need
![image](https://user-images.githubusercontent.com/24967648/70386786-dc763080-1951-11ea-94ad-d925e7a917b7.png)
![image](https://user-images.githubusercontent.com/24967648/70386798-062f5780-1952-11ea-8fa8-8dd2a43d2a36.png)

**Comment by philiphoyos** [comment]
I have the same issue, it seems that neither of the arguments specified in either the publish profile.xml, or the additional arguments is honored. I've tested with
`/BlockOnPossibleDataLoss=false /DropObjectsNotInSource=true /TreatVerificationErrorsAsWarnings=false`

**Comment by jeroenski74** [comment]
Hi,

I have changed is to true because we want people to think about it, I have not tested it again after. Should there be something changed?

Verzonden vanuit Mail<https://go.microsoft.com/fwlink/?LinkId=550986> voor Windows 10

Van: Eavanshi Arora<mailto:notifications@github.com>
Verzonden: donderdag 5 december 2019 06:19
Aan: microsoft/azure-pipelines-tasks<mailto:azure-pipelines-tasks@noreply.github.com>
CC: jeroenski74<mailto:jeroenski74@hotmail.com>; Mention<mailto:mention@noreply.github.com>
Onderwerp: Re: [microsoft/azure-pipelines-tasks] Azure SQL DacPac deployment failing when trying to drop a column even though /p:BlockOnPossibleDataLoss=false is set for SqlPackage.exe (#11191)


@jeroenski74<https://github.com/jeroenski74> Can you please confirm if you are still facing the issue? Thanks!

**Comment by eaarora-ms** [comment]
@Benjin  Can you please look into it?

**Comment by ronbuchanan** [comment]
Experiencing the same issue. /p:BlockOnPossibleDataLoss=false was not honored when trying to drop a column, but adding /p:TreatVerificationErrorsAsWarnings=true allowed the release to continue for one database, but not another. Looks like there is a major issue with it honoring parameters for these releases.

**Comment by eaarora-ms** [comment]
@ronbuchanan  @philiphoyos @sankara-narayanan Can any of you please confirm the following?

 1.	Is this against a DW target?
2.	Which version of Dacfx is used? The latest bits are available at - https://docs.microsoft.com/en-us/sql/tools/sqlpackage-download?view=sql-server-ver15 
3.	Is the “Data loss could occur” warning blocking the deployment or is there any error in script execution after that. Can you please try the script action with sqlpackage and check which step this is failing on.
4. Can you please share the full commandline used with sqlpackage - Is DropObjectsNotInSource/ TreatVerificationErrorsAsWarnings used?

**Comment by udeeshagautam** [comment]
Hi,

We are looking for a repro to see if this issue is in DacFx layer for investigation, but not able to repro this using latest SqlPackage. 
 
I understand that the deploy is failing when there is possible data loss as part of Azure DevOps pipeline even with /p:BlockOnPossibleDataLoss=false. Can someone please confirm that if latest SqlPackage.exe is used directly from command line for dacpac deployment with same /p:BlockOnPossibleDataLoss=false option - the same behavior still shows?

Please find the latest SqlPackage at https://docs.microsoft.com/en-us/sql/tools/sqlpackage-download?view=sql-server-ver15 

Thanks for your help.

**Comment by ronbuchanan** [comment]
@udeeshagautam I updated to the latest SqlPackage on our deploy agents and the error is still occurring. Unfortunately, my workaround using an XML profile file now breaks, too. I have no way to deploy a database now without manually deploying.

**Comment by udeeshagautam** [comment]
Thanks @ronbuchanan for trying with latest bits. Can you please confirm once that you are trying **directly deploying via SqlPackage.exe from the command line** OR through deploy agents. Requesting this clarification to understand if the paramter is reaching sqlpackage.exe and the result is passed as is or not. Please note that using sqlpackage directly from commandline is the key thing here to investigate if the issue is in SqlPackage.exe. Thanks again for the help!

**Comment by ronbuchanan** [comment]
@udeeshagautam to clarify...

* manually running SqlPackage.exe from the agent CLI deploys **successfully** using the publish profile file (copied and pasted what Azure DevOps executed)

`"C:\Program Files\Microsoft SQL Server\150\DAC\bin\SqlPackage.exe" /Action:Publish /SourceFile:"source.dacpac" /TargetServerName:"target_db" /TargetDatabaseName:"target_db" /TargetUser:"target_user" /TargetPassword:"********" /Profile:"source.publish.xml"  /TargetTimeout:120`

* running our pipeline in Azure DevOps fails, regardless of using arguments or a publish profile

It appears that the Pipeline is treating the warnings as errors and so fails the deploy.

**Comment by udeeshagautam** [comment]
Thanks for the Clarification. Will follow up on the same. Thanks again!

**Comment by Roshan-Kumar-Microsoft** [comment]
Not able to reproduce in our local . This is passing for me 
![SqlDacpacDeployWorkingFine](https://user-images.githubusercontent.com/57892372/72242559-c5d98e00-360f-11ea-9984-f49986b63ac7.PNG)

Can you please send me where we can reproduce this issue ?

**Comment by jeroenski74** [comment]
Hi,

We don't use it anymore and take care of the data upfront to prevent the error. Can't check it easily anymore.

Maybe in the future we will, but checking upfront is also more save so probably we will stick to it.

With kind regards,
Jeroen

Outlook voor Android downloaden<https://aka.ms/ghei36>

________________________________

**Comment by koder101** [comment]
I am also getting the same issue that `/p:BlockOnPossibleDataLoss=False` is not honored when running through Azure DevOps CI/CD, however, publishing from local works after explicitly unchecking the setting in Advanced.
Any ETA on this fix..?

**Comment by ronbuchanan** [comment]
![devops_snapshot](https://user-images.githubusercontent.com/729849/72543278-826f6180-3853-11ea-89fe-8558053a8f23.jpg)

@Roshan-Kumar-Microsoft I am not sure what other details you want for reproducing. This is the error I see, even with the proper parameters set in Azure DevOps and/or an XML deploy profile. I can run it from CLI using SqlPackage, but not from within a Pipeline.

**Comment by arjgupta** [comment]
@ronbuchanan can you try to run the sqlpackage command manually from a  powershell prompt. You should get a similar warning. After running the command, print the value of the variable $LASTEXITCODE. If it is non zero, then that means sqlpackage is executing with non zero exit code and we fail the task if sqlpackage exits with non zero exit code. Please let us know the results.

**Comment by GryphonsClaw** [comment]
We are having the same issue

**Comment by SumiranAgg** [comment]
Closing this issue. If you still see the issue please provide above information necessary to debug further and re-open the ticket.

**Comment by jacderida** [comment]
Sorry, but I don't think this should be closed. It's still an issue for me on the `SqlAzureDacpacDeployment` task. I have the following:
```
- task: SqlAzureDacpacDeployment@1
            displayName: 'deploy warehouse database changes'
            inputs:
              azureSubscription: '<snipped>'
              serverName: '$(AnalyticsWarehouseSqlServer)'
              databaseName: '$(AnalyticsWarehouseDbName)'
              sqlUsername: '$(AnalyticsWarehouseDbLogin)'
              sqlPassword: '$(AnalyticsWarehouseDbPassword)'
              dacpacFile: '$(system.artifactsDirectory)/data-warehouse-dacpac/CollectivWorks.DataWarehouse.dacpac'
              additionalArguments: >-
                /p:CommandTimeout="1200"
                /p:BlockOnPossibleDataLoss=False
                /v:PresenceFeedEnabled='0'
                /TargetTimeout:"1200"
```

My pipeline fails with:
```
##[error]*** Could not deploy package.
##[error]Warning SQL72015: The column [dbo].[CalendarDim].[MonYear] is being dropped, data loss could occur.
Warning SQL72015: The column [dbo].[CalendarDim].[YYYYMM] is being dropped, data loss could occur.
Warning SQL72015: The ty
##[error]pe for column MonthYear in table [dbo].[CalendarDim] is currently  VARCHAR (14) NULL but is being changed to  CHAR (7) NULL. Data loss could occur.
```

**Comment by ntiyiso** [comment]
I am still facing the same issue. I have tried overriding dacpac settings with a publish profile pre-configured to allow data loss and also tried passing /p:BlockOnPossibleDataLoss=false parameter to SQLPackage.exe in my release pipeline. However, the issue persists 

I've done further testing to see if the issue is with Azure SQL Database deployment task or something else, it seems like these affects Azure SQL database in general as the issue also appears in Visual Studio 2017 and 2019 publish feature

**Comment by Jetski5822** [comment]
Yeah im getting this too

```
2020-03-18T11:58:18.6603675Z Warning SQL72015: The type for column in_or_out in table [dbo].[txhr0config_ach_round] is currently  CHAR (1) NULL but is being changed to  BIT NULL. Data loss could occur.
2020-03-18T11:58:18.6603762Z Warning SQL72015: The column [dbo].[txhr0Visa].[Restricted] is being dropped, data loss could occur.
2020-03-18T11:58:18.6605681Z Error SQL72014: .Net SqlClient Data Provider: Msg 245, Level 16, State 1, Line 1 Conversion failed when converting the varchar value 'Y' to data type bit.
2020-03-18T11:58:18.6605808Z Error SQL72045: Script execution error.  The executed script:
2020-03-18T11:58:18.6605891Z ALTER TABLE [dbo].[txhr0config_ach_round] ALTER COLUMN [in_or_out] BIT NULL;
```

I dont care about the data in the table, so not sure why its erroring

**Comment by atmalone** [comment]
I am experiencing the exact same issue. 
Why is it blocking if I have `    <BlockOnPossibleDataLoss>False</BlockOnPossibleDataLoss>` set to false in my publish profile? I am altering a column and it makes releases fail.

Please update this.

`2020-04-07T11:09:35.2859817Z Starting rebuilding table [dbo].[Image]...
2020-04-07T11:09:38.7465816Z An error occurred while the batch was being executed.
2020-04-07T11:09:38.7552992Z Updating database (Failed)
2020-04-07T11:09:38.7812264Z *** Could not deploy package.
2020-04-07T11:09:38.7812927Z Warning SQL72015: The type for column KeyName in table [dbo].[Image] is currently  VARCHAR (250) NULL but is being changed to  VARCHAR (100) NULL. Data loss could occur.
2020-04-07T11:09:38.7813452Z Warning SQL72015: The type for column Name in table [dbo].[Image] is currently  VARCHAR (250) NULL but is being changed to  VARCHAR (100) NULL. Data loss could occur.
2020-04-07T11:09:38.7813869Z Error SQL72014: .Net SqlClient Data Provider: Msg 8152, Level 16, State 30, Line 36 String or binary data would be truncated.
2020-04-07T11:09:38.7816280Z Error SQL72045: Script execution error.  The executed script:
2020-04-07T11:09:38.7816541Z BEGIN TRANSACTION;
2020-04-07T11:09:38.7816677Z 
2020-04-07T11:09:38.7816961Z SET TRANSACTION ISOLATION LEVEL SERIALIZABLE;
2020-04-07T11:09:38.7817124Z 
2020-04-07T11:09:38.7817319Z SET XACT_ABORT ON;
2020-04-07T11:09:38.7817450Z 
2020-04-07T11:09:38.7817693Z CREATE TABLE [dbo].[tmp_ms_xx_Image] (
2020-04-07T11:09:38.7817947Z     [ImageId]          INT           IDENTITY (1, 1) NOT NULL,
2020-04-07T11:09:38.7818205Z     [Name]             VARCHAR (100) NULL,
2020-04-07T11:09:38.7818492Z     [Description]      VARCHAR (255) NULL,
2020-04-07T11:09:38.7818747Z     [Tags]             VARCHAR (100) NULL,
2020-04-07T11:09:38.7818993Z     [Location]         VARCHAR (255) NULL,
2020-04-07T11:09:38.7819268Z     [CreatedBy]        INT           NOT NULL,
2020-04-07T11:09:38.7819552Z     [CreatedOn]        DATETIME      NOT NULL,
2020-04-07T11:09:38.7819801Z     [UpdatedBy]        INT           NULL,
2020-04-07T11:09:38.7820091Z     [UpdatedOn]        DATETIME      NULL,
2020-04-07T11:09:38.7820343Z     [Version]          INT           NOT NULL,
2020-04-07T11:09:38.7820624Z     [KeyName]          VARCHAR (100) NULL,
2020-04-07T11:09:38.7820906Z     [ConvertToJpg]     BIT           NULL,
2020-04-07T11:09:38.7821152Z     [Width]            INT           NULL,
2020-04-07T11:09:38.7821400Z     [Height]           INT           NULL,
2020-04-07T11:09:38.7821670Z     [Resolution]       FLOAT (53)    NULL,
2020-04-07T11:09:38.7821919Z     [Timestamp]        ROWVERSION    NOT NULL,
2020-04-07T11:09:38.7822167Z     [CoordinateX]      FLOAT (53)    NULL,
2020-04-07T11:09:38.7822431Z     [CoordinateY]      FLOAT (53)    NULL,
2020-04-07T11:09:38.7822695Z     [CoordinateS]      FLOAT (53)    NULL,
2020-04-07T11:09:38.7822914Z     [fkTriggerState
2020-04-07T11:09:38.7823023Z 
2020-04-07T11:09:38.8381179Z ==============================================================================
2020-04-07T11:09:39.1417148Z ##[error]Error: Error executing C:\Program Files (x86)\Microsoft Visual Studio\2017\BuildTools\Common7\IDE\Extensions\Microsoft\SQLDB\DAC\150\sqlpackage.exe`

**Comment by motorreiziger** [comment]
Same Issue here with yaml sqldacpactask
##[error]is being dropped, data loss could occur.

**Comment by ronbuchanan** [comment]
Apologies for the long wait for my reply, but I have not changed our DevOps pipeline back to the DACPAC deploy. I am still using a Powershell script to execute SqlPackage.exe with all of the correct parameters. I can provide you with the output, but I'm not sure what good that would do since it is successful.

I will try to get to this test today and post the $LASTEXITCODE. 

Example CLI:
```
"C:\Program` Files\Microsoft SQL Server\150\DAC\bin\SqlPackage.exe" `
    /Action:Publish `
    /SourceFile:"source.dacpac" `
    /TargetServerName:"target_srvr" `
    /TargetDatabaseName:"target_db" `
    /TargetUser:"target_user" `
    /TargetPassword:"********" `
    /Profile:"source.publish.xml" `
    /TargetTimeout:120
```

**Comment by mez12345** [comment]
Is there any update on this. I am getting the same behaviour when using DevOps Release Pipelines

**Comment by ronbuchanan** [comment]
Here are the results after printing the value for `$LASTEXITCODE`, as requested.

NOTES:
- `***` is a redacted value, not wildcards

```
2020-05-11T18:06:10.4438119Z ##[section]Starting: Deploy databases
2020-05-11T18:06:10.4544547Z ==============================================================================
2020-05-11T18:06:10.4545025Z Task         : PowerShell
2020-05-11T18:06:10.4545461Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2020-05-11T18:06:10.4545862Z Version      : 2.165.0
2020-05-11T18:06:10.4546214Z Author       : Microsoft Corporation
2020-05-11T18:06:10.4546710Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2020-05-11T18:06:10.4547249Z ==============================================================================
2020-05-11T18:06:11.5858313Z Generating script.
2020-05-11T18:06:11.5869121Z Formatted command: . 'c:\agent\temp\***.ps1' "***" "***" "***" "c:\agent\temp\***.dacpac" "c:\agent\temp\***.publish.xml"
2020-05-11T18:06:11.6273227Z ========================== Starting Command Output ===========================
2020-05-11T18:06:11.6532883Z ##[command]"pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agent\_work\_temp\***.ps1'"
2020-05-11T18:06:12.1745257Z ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
2020-05-11T18:06:12.1753838Z Database Deployment
2020-05-11T18:06:12.1758993Z 
2020-05-11T18:06:12.1766940Z Parameters
2020-05-11T18:06:12.1774715Z ----------
2020-05-11T18:06:12.1782293Z Server:  ***
2020-05-11T18:06:12.1790239Z DACPAC:  c:\agent\temp\***.dacpac
2020-05-11T18:06:12.1797063Z Profile: c:\agent\temp\***.publish.xml
2020-05-11T18:06:12.1803281Z . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . . .
2020-05-11T18:06:12.1807101Z 
2020-05-11T18:06:12.3779685Z Started deploying database Database1 ...
2020-05-11T18:06:12.3805277Z Started deploying database Database2 ...
2020-05-11T18:06:12.3814978Z Started deploying database Database3 ...
2020-05-11T18:06:48.5256348Z Finished deploying database Database2 with exit code (0)
2020-05-11T18:06:51.0101914Z Finished deploying database Database3 with exit code (0)
2020-05-11T18:07:36.2344296Z Finished deploying database Database1 with exit code (0)
2020-05-11T18:07:36.2350332Z 
2020-05-11T18:07:36.2360433Z Database deployment completed
2020-05-11T18:07:36.2369694Z ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~ ~
2020-05-11T18:07:36.3296778Z ##[section]Finishing: Deploy databases
```

I will most likely continue to use my PS script because it allows me to deploy all of my databases in parallel.

**Comment by eaarora-ms** [comment]
Yes, SQLPackage.exe should have ignored the warning when /BlockOnPossibleDataLoss is being set to false. I will check it.

Meanwhile, can you please check in your DB project properties ( project > properties > debug) that checkbox for "Block on incremental deployment if data loss occurs" is disabled.

Screenshot

https://user-images.githubusercontent.com/16772764/41333279-e0828ede-6efe-11e8-9a8b-f761cf4ec9b4.png

**Comment by joeyeng** [comment]
@eaarora-ms Is there an ETA for a fix?

**Comment by asranja** [comment]
@udeeshagautam We are able to reproduce this outside of Azure DevOps as well. Using /BlockOnPossibleDataLoss=false argument doesn't ignore the error. Can we have someone from the SqlPackage team help us with the issue.

**Comment by kincho-guerrero** [comment]
Same error here, not honoring the /p:BlockOnPossibleDataLoss=false inside of the pipeline. Besides the CLI task to deploy directly through command line, are there any workarounds for this issue?

**Comment by devops-lucina** [comment]
@kincho-guerrero We are just running a Powershell task that runs sqlpackage.exe with the right parameters. This works for us using our own build agents. Look above in my comments for more details.

**Comment by ronbuchanan** [comment]
@kincho-guerrero We are just running a Powershell task that runs sqlpackage.exe with the right parameters. This works for us using our own build agents. Look above in my comments for more details.

**Comment by itzkumar** [comment]
@ronbuchanan,

Can you please share us the PS script sample?

**Comment by ronbuchanan** [comment]
@itzkumar Here's the gist. The parameters are provided by the AzDO release.

https://gist.github.com/ronbuchanan/459dd77faa486a02520af4ef50d3c9e6

Hope it helps.

**Comment by cpujuice** [comment]
We were able to reproduce the issue by changing the data type of a column. Deployment fails when the underlying data cannot be directly mapped to the new data type, regardless of what BlockOnPossibleDataLoss is set to.

We are working on a fix for the next SqlPackage release.

**Comment by MichaelHolmesWP** [comment]
@ronbuchanan Can you confirm what version of sqlpackage.exe (ie the executable File/Product Version) your build agents are running?

**Comment by ronbuchanan** [comment]
@MichaelHolmesWP 
![image](https://user-images.githubusercontent.com/729849/85475798-aebaa700-b584-11ea-9513-422317a9e1c6.png)

**Comment by cpujuice** [comment]
Hello,

After further investigation, the errors we were able to reproduce are actually thrown by the SQL engine and not by SqlPackage itself. Currently SqlPackage only supports implicit conversion between data types. The BlockOnPossibleDataLoss option is honored when the deployment plan generates in the background, but when it executes it could still fail if the underlying data cannot be directly mapped to the new data type (hence errors like "String or binary data would be truncated" or "Arithmetic overflow error for data type").

For this we plan on updating the warning messages to be clearer that even though possible data loss is treated as a warning, any data requiring explicit conversion could still cause deployment to fail.

**Comment by cpujuice** [comment]
We are still not able to reproduce the issue with dropping columns though. Column is dropped successfully with BlockOnPossibleDataLoss=False regardless if there's data or not. Any pointers on this would be greatly appreciated!

**Comment by tcunbeliever** [comment]
I have resolved both of these issues with the creation of  pre-deployment sql scripts.  

For columns getting dropped - check for existence of the column then if present issue the alter table drop column command.

For columns changing from null to not null - check for the existence of the table (so new environments won't fail on the update) then if present issue an update command to change any null values to a default value.

After putting these in I was able to get deployments working, though it's a bit of extra scripting and totally a work around since really I think the system should do these things itself (it just doesn't work correctly).

**Comment by oscarelo** [comment]
Hi @tcunbeliever please can you share your pre-deployment sql scripts?

**Comment by bmvij** [comment]
Seems like it doesn't work from Publish
Go to  DB project -- properties ( project > properties > debug) uncheck "Block on incremental deployment if data loss occurs" this will work. 

I am getting timeout error below, i have increased the timeout time but no luck. ( I am deploying it through PS)

    + CategoryInfo          : InvalidOperation: (:) [Invoke-Sqlcmd], PipelineStoppedException
    + FullyQualifiedErrorId : SqlExectionError,Microsoft.SqlServer.Management.PowerShell.GetScriptCommand
 
Invoke-Sqlcmd : Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.

**Comment by adamsclark** [comment]
Is there any update on this. I am getting the same behavior when using DevOps Release Pipelines

**Comment by oscarelo** [comment]
After many tries, I was able to make it work!

In the Additional Arguments for the Release Pipeline put:
/p:BlockOnPossibleDataLoss=false /p:DropObjectsNotInSource=true

![image](https://user-images.githubusercontent.com/65922847/89959632-835a4d00-dc02-11ea-907d-c24a4b5cea31.png)

If you still are getting the error, you need to be sure that all the referenced databases, tables, columns, etc, exists in the target.
For example, if you have a store procedure with a reference to a table or column from a different database and the column is not there you will have this problem and you will get a lot of error messages for other things not directly related with the specific problem, so this can be very confusing. Sometimes you don't need to pay to much attention to the firsts errors in the list and just check the one that you have at the end and fix it adding the missing things in the target server.

On the other hand, if you have views with references to other tables or columns from different databases, in this case (Views) you will need to add the datapac as a reference in the project, to do this you need to create a Database project for the other referenced database, do build and rebuilt so you can generate the dacpac file in the repo folder C:\Users\userLogin\source\repos\Solution\Project\bin\Debug, then you can use the generated dacpac to add it in the Visual Studio project as a reference.

![image](https://user-images.githubusercontent.com/65922847/89961778-64f75000-dc08-11ea-930a-95bfb88237d0.png)

I hope this helps somebody.

Good luck!

**Comment by VeerabhadramRagi** [comment]
Hi, 
Could you please look into the below issue:
Facing the below issue in Release pipeline (it is defined for Azure SQL DB):
_**Warning SQL72012: The object [data_0] exists in the target, but it will not be dropped even though you selected the 'Generate drop statements for objects that are in the target database but that are not in the source' check box**_
Please help me to fix it.

Thanks

**Comment by 20shivangi** [comment]
Is the issue still occuring with /p:BlockOnPossibleDataLoss=false ?
Or should we aks sql team to take a look into it

**Comment by bmvij** [comment]
Yes

**Comment by sven73a** [comment]
yes issue is still there

**Comment by adamsclark** [comment]
Issue still happens for me as well with `/p:BlockOnPossibleDataLoss=false`

**Comment by 20shivangi** [comment]
Is this issue also occuring when you are using hosted agents ? Please share the full debug logs with variable system.debug to true in your pipeline/release.

I tried to repro the issue but was unable to do the same

**Comment by ankitsatapara07** [comment]
yes issue is still there

/p:BlockOnPossibleDataLoss=false is not honored.

Some logs:
```
##[error]Warning SQL72015: The column [XXX].[YYYYYYY].[ZZZZZZZZ] is being dropped, data loss could occur.
##[error]ableId] is being dropped, data loss could occur.
```

**Comment by ankitsatapara07** [comment]
Finally Figure it out what causing the issues.

Actually problem is not with `/p:BlockOnPossibleDataLoss=false` , It is being honored but the Drop Blocking Assemblies are not allowed to drop the column(s)/table(s).

To work it perfectly as you want use this flag : `/p:AllowDropBlockingAssemblies=true` and will work as expected.

I don't think so issue is within the azure pipeline.

JFYI - I am using the hosted agent vs2017.

**Comment by denis-peshkov** [comment]
I can approve the issue is not fixed

`/p:IncludeCompositeObjects=True /p:ScriptDatabaseOptions=False /p:BlockOnPossibleDataLoss=False /p:AllowDropBlockingAssemblies=true /p:DisableAndReenableDdlTriggers=False /p:DoNotAlterReplicatedObjects=False /p:DoNotAlterChangeDataCaptureObjects=False /p:DropConstraintsNotInSource=False  /p:DropIndexesNotInSource=False  /p:DropExtendedPropertiesNotInSource=False  /p:DropDmlTriggersNotInSource=False /p:ExcludeObjectTypes=ExternalDataSources;Users;UserDefinedTableTypes;UserDefinedDataTypes;Credentials;DatabaseScopedCredentials /p:CommandTimeout=3600 /p:IgnoreColumnOrder=True`

error:
```ps
*** Verification of the deployment plan failed.
Warning SQL72015: The column [dbo].[virtinv_Invoice].[CreatedBy] is being dropped, data loss could occur.
Warning SQL72015: The column [dbo].[virtinv_Invoice].[CreatedOn] is being dropped,
2 more errors. Click on expand view in the context menu to view complete logs.
```

I have included both:
`/p:BlockOnPossibleDataLoss=false` and `/p:AllowDropBlockingAssemblies=true `

**Comment by jtbrown52** [comment]
We had this same issue and thank you @denis-peshkov for your update - it helped! The deployment still failed, however, it actually performed what we needed it to do - and thinking about it, there are a couple of things we could have done to help things along more smoothly.

In our particular situation, we had a  few contributing factors that may have led to this dilemma.

- a column that was actually being REMOVED from a table that had foreign key constraints as well as object inter-dependencies, however, those objects were actually being dropped from the database as well as part of the release.  

- In addition we dropped a few tables and views in our code as they were no longer needed, however, I think the way they were dropped was partly at issue (dropped in a dev database and re-synced to the project rather than dropped out of the project itself to allow the refactor log to be properly updated).

- We had several tables marked as temporal tables that were being changed as part of the deployment.  My observation on dacpacs with temporal tables is that its a bit "brittle" and probably requires handling before-hand in a pre-deployment script.

My main takeaway here is that if you're dropping objects from your schema with dependencies and existing data, you're definitely going to want to handle that in a pre-deployment script.  Likely what you're going to have to do is script out disabling any temporal properties on tables you're altering (which you should be able to script out using visual studio), and use the schema compare to handle any changes to tables for which you're dropping any columns (or at least handle the data conversion/DML).   Then if you use /p:BlockOnPossibleDataLoss=false /p:AllowDropBlockingAssemblies=true  /p:TreatVerificationErrorsAsWarnings=true you should be good to go, just be aware that you're ok with losing data in the objects you're dropping (backup always).

**Comment by davidhwang-ij** [comment]
My release failed even with these three parameters:
/p:BlockOnPossibleDataLoss=false, /p:AllowDropBlockingAssemblies=true , /p:TreatVerificationErrorsAsWarnings=true

BUT I resolved the issue by manually updating the Azure SQL within SSMS with the updated database first.
- Right click the local database > Tasks > Export Data-tier Application (save in a local folder)
- Make a connection to Azure SQL using your server name <name>.database.windows.net in your azure portal.
- Delete the original table, right click the Databases folder > Tasks > Import Data-tier Application
- Run the pipeline again

Hope that helped.

**Comment by Anunaya02** [comment]
I am using the same parameters in at least 10 database project and it is working fine for me. I am using DevOps pipeline for deploying my database on On-Prem as well Azure. 

Can anybody please specify what steps I can take to replicate the issue? 

An easy way to debug your issue is to use Publish and Generate script option and you will be able to see what script it is actually deploying in backend.

I tried the changes on VS 2017 and VS2019. SSDT version is 16.0.62010.06180

**Comment by dzsquared** [comment]
Hi folks, we have clarified the property "BlockOnPossibleDataLoss" in the SqlPackage documentation as a deployment will still exit/fail if actual data exists on the database that will be impacted by changes when the property is set to false.
https://docs.microsoft.com/en-us/sql/tools/sqlpackage/sqlpackage-publish?view=sql-server-ver15#properties-specific-to-the-publish-action

> “Specifies that the operation will be terminated during the schema validation step if the resulting schema changes could incur a loss of data, including due to data precision reduction or a data type change that requires a cast operation. The default (True) value causes the operation to terminate regardless if the target database contains data. An execution with a False value for BlockOnPossibleDataLoss can still fail during deployment plan execution if data is present on the target that cannot be converted to the new column type.”

As a few other users have observed, you may have interdependencies involved in the deployment that need to be handled by a predeployment script or an additional property such as p:DropObjectsNotInSource or p:AllowDropBlockingAssemblies may cover your scenario.

**Comment by 20shivangi** [comment]
Closing the issue as item as users with comments from the past year have been able to sort through their specific deployment scenario, the same got confirmed by @dzsquared on mail :)

**Comment by segunak** [comment]
> I have resolved both of these issues with the creation of pre-deployment sql scripts.
> 
> For columns getting dropped - check for existence of the column then if present issue the alter table drop column command.
> 
> For columns changing from null to not null - check for the existence of the table (so new environments won't fail on the update) then if present issue an update command to change any null values to a default value.
> 
> After putting these in I was able to get deployments working, though it's a bit of extra scripting and totally a work around since really I think the system should do these things itself (it just doesn't work correctly).

Would you be able to share your pre-deployment scripts? Seeing this same issue using `p:BlockOnPossibleDataLoss = false`.

Same as other commenters, Azure DevOps pipeline deployment failing when attempting to drop columns from an existing table.

**Comment by alicianoll** [comment]
I have been waiting patiently for this bug to be resolved but it persists. 
I do not believe the flag for BlockOnPossibleDataLoss will apply. 
THe issue I am still dealing with is when deleting a temporal table entirely. Fortunately, this isn't a frequent occurrence, and I have to remove the table manually out outside of the pipeline.
Has anyone else found this issue and an automated solution?
To reproduce this issue, should be to create a temporal table, delete it , and use SqlPackage.exe to deploy (I use devops release to execute). The problem is clear and visible when you save the generated script (to file) . I can see that any column changes I make to a temporal table are correctly handled (checks for temporal table state (temporal_type = 2) and ALTER's table turning SYSTEM_VERSIOING =OFF prior to attempting the change. However, with the DELETE, no checks or altering is done, rather it just attempts to delete the table, which fails

**Comment by segunak** [comment]
> I have been waiting patiently for this bug to be resolved but it persists. I do not believe the flag for BlockOnPossibleDataLoss will apply. THe issue I am still dealing with is when deleting a temporal table entirely. Fortunately, this isn't a frequent occurrence, and I have to remove the table manually out outside of the pipeline. Has anyone else found this issue and an automated solution? To reproduce this issue, should be to create a temporal table, delete it , and use SqlPackage.exe to deploy (I use devops release to execute). The problem is clear and visible when you save the generated script (to file) . I can see that any column changes I make to a temporal table are correctly handled (checks for temporal table state (temporal_type = 2) and ALTER's table turning SYSTEM_VERSIOING =OFF prior to attempting the change. However, with the DELETE, no checks or altering is done, rather it just attempts to delete the table, which fails. If anyone has found a clever solution, I would really appreciate it :) Thanks

Disclaimer: What follows is a workaround and not a "real" solution. It's basically saying `handle drops before you run SQLPackage so that by the time you do, you know it'll work`. I wrote some more about this approach in [this article](https://segunakinyemi.com/blog/dacpac-dropping-objects/) and also what I've found online about why SQLPackage doesn't handle certain drops well.

What we've had to do for the project I'm working on is handle dropping tables, views, columns, and objects in general manually via what's we're calling a "pre-pre-deployment" script to work around not being able to do so via SQLPackage. Our process is below.

1. Engineer realizes they need to drop something from the project. We're using a [SQL Server Database Project](https://learn.microsoft.com/en-us/sql/ssdt/how-to-create-a-new-database-project) to manage changes.
2. Engineer tries to do so the "normal" way. Delete the table (or whatever object) from the database project via Visual Studio and then run our deployment, which uses SQLPackage (called from an Azure DevOps pipeline), to one of our pre-production environments. This is basically testing if SQLPackage is going to play nice. We have `BlockOnPossibleDataLoss=False` and `DropObjectsNotInSource=True` configured on all deployments, and yet sometimes we get failed deployments when trying to drop things, which is what led me to this thread.
3. If the deployment passes pre-production, we go forward with the change that will run via SQLPackage. No need for shenanigan's. End here.
4. If the deployment is failing as seen in this issue, we handle whatever object drops/manipulations we need to do via a script that runs before the SQLPackage deployment (pre-pre-deployment) script.

So, what you said about removing the table manually outside the pipeline? Yeah, that's the solution. We just automated that part so that there's a script in our repository that always runs before the SQLPackage, and if we ever need to drop something that SQLPackage will cry about, we do it there. 

```mermaid
graph LR
step1["Try SQLPackage in pre-prod"] --> step2{"Did it work?"}
step2 --> |yes| worked["Trust in SQLPackage"]
step2 --> |no| notWorked["Put drops in<br/>pre-pre-deployment script"]
worked --> fd["Final Deployment"]
notWorked --> fd
```

The pre-pre-deployment script would need to have commands written conditionally so that they only do something if necessary.

```SQL
IF EXISTS (SELECT * FROM sys.tables WHERE name ='SomeTemporalTable') 
BEGIN
  ALTER TABLE SomeTemporalTable SET (SYSTEM_VERSIONING = OFF);
  DROP TABLE SomeTemporalTable
  DROP TABLE IF EXISTS SomeTemporalTableHistory
END
```

During a first deployment with the script above `SomeTemporalTable` and it's history table would get dropped. But on the next deployment, it'd do nothing, so the script itself would be "safe" and you wouldn't have to worry that it would cause failures on future deployments by trying to mutate objects that no longer exist. 

Not sure if this helps your use case though. 

>  To reproduce this issue, should be to create a temporal table, delete it , and use SqlPackage.exe to deploy (I use devops release to execute).

I'm assuming the "delete it" part is something that you did manually on the database, and even afterwards, SQLPackage deployments are having phantom references to it? Or is that you deleted it with the intention that SQLPackage would re-create it on the next deployment? Your scenario might be different than what than the one I'm encountering where we want to delete an object (table, view, column, etc.) permanently.
